### PR TITLE
inferCumulativity: do not unfold constants unless necessary

### LINKED ISSUE #11741: Cumulativity checking(?) extremely slow
#### Description of the problem
I keep running into this problem and it seems I haven't actually opened an issue for it so here goes:
enabling cumulativity can slow down the checking(?) of a record a whole lot.
Here's a minimized example program:

https://gist.github.com/Janno/4545b80f8769d839c39ae01903992468

The final `Record` command still hasn't finished on my machine after ~10 minutes. Making the record `NonCumulative` makes it go through in the blink of an eye.

There was some discussion about this problem on gitter a while ago.
https://gitter.im/coq/coq?at=5e21cdb43ea53f0f663d6de3

#### Coq Version
master

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fix #11741

### CODE REVIEW

**Comment by ppedrot** [comment]
This seems to change the universe heuristics for HoTT...

**Comment by ppedrot** [comment]
@coqbot ci minimize

**Comment by Alizter** [comment]
Easy peasy fix, 
```coq
    first [apply ishprop_ooextendable@{a a i i i
                                       i i i i i}
           | apply ishprop_ooextendable@{a a i i i
                                       i i i i i
                                       i i i}
           | apply ishprop_ooextendable@{a a i i i
                                         i i i i i
                                         i i i i i}].
```
This was already a cross-version overlay tactic. The number of universes going down is always welcome.

**Comment by SkySkimmer** [comment]
>Easy peasy fix,

That's a misunderstanding of what's going on, the problem is that some inductives declared Cumulative got inferred noncumulative.

**Comment by SkySkimmer** [comment]
See https://github.com/SkySkimmer/HoTT/commit/9cf729fa80b737b2ad477f99705db28887ce2cac for a real fix
However maybe we should instead have some other heuristic for constant unfolding.

**Comment by ppedrot** [comment]
I think that the heuristic presented in this PR is very reasonable. It is easy to explain and predict and solves the blowup problem. I am quite in favour of merging this (with overlays and changelog and doc, but that's not the design part).

**Comment by SkySkimmer** [comment]
Here is a fancier heuristic, which is fast enough on https://github.com/coq/coq/issues/11741 yet produces optimal results. See also some info in https://github.com/coq/coq/issues/11741#issuecomment-1093225153

**Comment by SkySkimmer** [comment]
It's allowing 
~~~coq
Set Universe Polymorphism.
Cumulative Inductive foo@{*i}  (T:=Type@{i}:Type@{i+1}) (A:Type@{i}) : Type@{i+1}
  := bar (X:T) : foo A.
~~~
which is inconsistent (`i` is covariant not irrelevant, see also #11039).
(with both heuristics, the issue is too little unfolding)

I guess we need to always unfold rels,vars and section definitions, or set the universes they mention to invariant?

**Comment by SkySkimmer** [comment]
See also https://github.com/coq/coq/issues/15916

**Comment by ppedrot** [comment]
@SkySkimmer I've observed that in my PR for template poly before. Isn't this a current bug of soundness in the cumulativity computation actually?

**Comment by ppedrot** [comment]
@SkySkimmer can you rebase?

**Comment by SkySkimmer** [comment]
Should be ready

File: kernel/inferCumulativity.ml [hunk 1]
```diff
@@ -172,9 +170,27 @@ let rec infer_fterm cv_pb infos variances hd stk =
...
-    infer_stack infos variances stk
+  | FFlex Names.(RelKey _ | VarKey _ as fl) ->
+    (* We could try to lazily unfold but then we have to analyse the
+       universes in the bodies, not worth coding at least for now. *)
+    begin match unfold_ref_with_args (fst infos) (snd infos) fl stk with
+    | Some (hd,stk) -> infer_fterm cv_pb infos variances hd stk
+    | None -> infer_stack infos variances stk
+    end
+  | FFlex (Names.ConstKey con as fl) -> begin
+      let def = unfold_ref_with_args (fst infos) (snd infos) fl stk in
```
Comment (ppedrot) [inline kernel/inferCumulativity.ml:181]: Nitpick: the indentation is not consistent with the other branches.

Comment (ppedrot) [reply kernel/inferCumulativity.ml:181]: (I.e. move the begin on its own line.)

Comment (SkySkimmer) [reply kernel/inferCumulativity.ml:181]: Moving the begin doesn't change indentation.
Do you want me to indent less than the automation suggests? I can do that.

File: kernel/inferCumulativity.ml [hunk 2]
```diff
@@ -172,9 +170,27 @@ let rec infer_fterm cv_pb infos variances hd stk =
...
+      try
+        try
+        let ov = variances in
+        let variances = infer_constant (info_env (fst infos)) variances con in
+        let variances = infer_stack infos variances stk in
+        if Option.has_some def && variances != ov then raise Unfold else variances
+        with TrivialVariance when Option.has_some def -> raise Unfold
+      with BadVariance _ | Unfold as e ->
```
Comment (SkySkimmer) [inline kernel/inferCumulativity.ml:189]: I remember thinking that there should be a nicer way to make this work than this exception system but I don't remember what that way was supposed to be.

File: kernel/inferCumulativity.ml (hunk 1 above)
Comment (ppedrot) [reply kernel/inferCumulativity.ml:181]: No, I mean leave the block alone and just move the begin on its own line with less indentation than the inner block.

Comment (SkySkimmer) [reply kernel/inferCumulativity.ml:192]: I don't see the point.

Comment (ppedrot) [reply kernel/inferCumulativity.ml:192]: The current code slightly confused me because I didn't match the end block with it's opening begin immediately.

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
CI failures unrelated. @coqbot merge now

**Pull request merged**